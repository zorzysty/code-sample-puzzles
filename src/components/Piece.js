import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import imgSrc from '../assets/picture.jpg';
import { DIMENSIONS } from '../const';


const spec = {
  beginDrag: (props) => {
    props.handleDrag();
    return props.piece;
  },
  endDrag: (props, monitor) => {
    // return piece to original position when dropped outside of target
    if (!monitor.didDrop()) return;

    // pass piece to handleDrop method if placed correctly and false otherwise
    props.handleDrop(monitor.getDropResult().success && props.piece);
  },
};

function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  };
}

class Piece extends PureComponent {
  render() {
    const {
      isDragging, // eslint-disable-line react/prop-types
      connectDragSource, // eslint-disable-line react/prop-types
      style,
      piece,
    } = this.props;

    // hide piece in drawer while it's being dragged
    const opacity = isDragging ? 0 : 1;
    // prevent overlaying of images while dragging
    const zIndex = isDragging ? 0 : 10;
    // set proper piece image using CSS sprites
    const left = -(piece.id % DIMENSIONS.columns) * DIMENSIONS.pieceSize;
    const top = -Math.floor(piece.id / DIMENSIONS.rows) * DIMENSIONS.pieceSize;

    return connectDragSource(
      <div
        className="Piece"
        style={{
          background: `url(${imgSrc})`,
          backgroundPosition: `${left}px ${top}px`,
          backgroundSize: `${DIMENSIONS.columns * 100}%`,
          opacity,
          zIndex,
          ...style,
        }}
      />,
    );
  }
}

export default DragSource('piece', spec, collect)(Piece);

Piece.propTypes = {
  piece: PropTypes.shape({
    id: PropTypes.number.isRequired,
  }).isRequired,
  handleDrop: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  handleDrag: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  style: PropTypes.objectOf(PropTypes.string),
};

Piece.defaultProps = {
  handleDrop: () => {},
  handleDrag: () => {},
  style: {},
};
