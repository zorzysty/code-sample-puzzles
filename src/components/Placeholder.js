import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';


function collect(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    hovered: monitor.isOver(),
  };
}

const spec = {
  drop: (props, monitor, component) => ({
    // set success prop of dropResult depending on whether piece was placed in correct place
    success: monitor.getItem().id === component.props.index,
  }),
};

class Placeholder extends PureComponent {
  render() {
    const { connectDropTarget, hovered } = this.props;

    return connectDropTarget(
      <div className={`Placeholder ${hovered ? 'Placeholder-hovered' : ''}`} />,
    );
  }
}

Placeholder.propTypes = {
  connectDropTarget: PropTypes.func.isRequired,
  hovered: PropTypes.bool.isRequired,
  index: PropTypes.number.isRequired, // eslint-disable-line react/no-unused-prop-types
};

export default DropTarget('piece', spec, collect)(Placeholder);
