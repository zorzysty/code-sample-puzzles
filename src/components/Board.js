import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Placeholder from './Placeholder';
import Piece from './Piece';
import { DIMENSIONS } from '../const';


class Board extends PureComponent {
  constructor(props) {
    super(props);
    this.pieceNumbers = [...Array(DIMENSIONS.pieceCount).keys()];
  }

  drawPieces() {
    const { pieces } = this.props;

    // render Piece for every piece in pieces array and a placeholder (drop target) for the rest
    return this.pieceNumbers.map((number) => {
      const piece = pieces.find(({ id }) => id === number);
      return piece
        ? <Piece key={number} piece={piece} />
        : <Placeholder key={number} index={number} />;
    });
  }

  render() {
    return (
      <div className="Board">
        {this.drawPieces()}
      </div>
    );
  }
}

export default Board;

Board.propTypes = {
  pieces: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.number })).isRequired,
};
