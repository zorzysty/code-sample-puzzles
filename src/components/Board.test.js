import React from 'react';
import { shallow } from 'enzyme';
import Board from './Board';
import Piece from './Piece';
import Placeholder from './Placeholder';


describe('Board', () => {
  const pieces = [
    { id: 0 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 6 }, { id: 7 }, { id: 8 },
  ];

  const missingPieces = [
    { id: 1 }, { id: 5 },
  ];

  const component = shallow(<Board pieces={pieces} />);

  it('should render correctly with pieces', () => {
    expect(component)
      .toMatchSnapshot();
  });

  it('should render correct Piece components', () => {
    expect(component
      .containsAllMatchingElements(pieces.map(piece => <Piece piece={piece} />)))
      .toBe(true);

    expect(component
      .containsAnyMatchingElements(missingPieces.map(piece => <Piece piece={piece} />)))
      .toBe(false);
  });

  it('should render correct Placeholder components', () => {
    expect(component
      .containsAllMatchingElements(missingPieces.map(piece => <Placeholder index={piece.id} />)))
      .toBe(true);

    expect(component
      .containsAnyMatchingElements(pieces.map(piece => <Placeholder index={piece.id} />)))
      .toBe(false);
  });

  it('should render correctly without pieces', () => {
    const emptyComponent = shallow(<Board pieces={[]} />);

    expect(emptyComponent.containsMatchingElement(<Piece />))
      .toBe(false);

    expect(emptyComponent)
      .toMatchSnapshot();
  });
});
