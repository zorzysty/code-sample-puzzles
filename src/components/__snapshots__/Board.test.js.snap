// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render correctly with pieces 1`] = `ShallowWrapper {}`;

exports[`Board should render correctly without pieces 1`] = `ShallowWrapper {}`;
